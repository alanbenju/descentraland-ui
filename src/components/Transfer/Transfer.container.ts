import { connect } from 'react-redux'
import { transferRequest } from '../../modules/transaction/actions'
import {
  getError,
  isTransfering,
} from '../../modules/transaction/selectors'
import { RootState } from '../../modules/types'
import { MapDispatch, MapDispatchProps, MapStateProps } from './Transfer.types'
import Transfer from './Transfer'
import { isConnected } from '../../modules/wallet/selectors'

const mapState = (state: RootState): MapStateProps => ({
  isConnected: isConnected(state),
  isTransfering: isTransfering(state),
  error: getError(state)
})

const mapDispatch = (dispatch: MapDispatch): MapDispatchProps => ({
  onTransfer: (addressToTransfer: string, amountToTransfer: number) => dispatch(transferRequest(addressToTransfer, amountToTransfer)),
})

export default connect(mapState, mapDispatch)(Transfer)
