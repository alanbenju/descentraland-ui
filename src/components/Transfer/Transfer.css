
  .Transfer .error {
    margin-top: 20px;
    color: var(--primary);
  }

  .Transfer .ui.modal .header{
    margin-bottom: 0px;
  }
  
  .subtitle{
    margin-bottom: 20px;
    font-size: 16px;
  }