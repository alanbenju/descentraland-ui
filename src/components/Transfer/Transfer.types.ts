import { AnyAction, Dispatch } from 'redux'
import { TransferRequestAction } from '../../modules/transaction/actions'
import { ConnectWalletRequestAction } from '../../modules/wallet/actions'

export type Props = {
  isConnected: boolean
  error: string | null
  onTransfer: (addressToTransfer: string, amountToTransfer: number) => void
  isTransfering: boolean
}

export type MapStateProps = Pick<
  Props,
  'isConnected' | 'error' | 'isTransfering'
>
export type MapDispatchProps = Pick<Props, 'onTransfer'>
export type MapDispatch = Dispatch<TransferRequestAction | ConnectWalletRequestAction | AnyAction>