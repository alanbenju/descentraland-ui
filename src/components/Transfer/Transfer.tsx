import React, { useEffect, useState } from 'react'
import {
    Button,
    Close,
    Field,
    Form,
    Header,
    Modal,
} from 'decentraland-ui'
import { Props } from './Transfer.types'
import './Transfer.css'

import { useNavigate } from "react-router-dom";

const Transfer: React.FC<Props> = ({
    onTransfer,
    isTransfering,
    error,
    isConnected
}) => {

    const [address, setAddress] = useState('');
    const [amount, setAmount] = useState(0);
    const [open, setOpen] = useState(true);

    const onSend = () => {
        console.log("ON SEND", address, amount)
        if (address && amount) onTransfer(address, amount)
    }
    
    let navigate = useNavigate();

    let back = (e: { stopPropagation: () => void; }) => {
        e.stopPropagation();
        setOpen(false)
        navigate(-1);
    };

    if (!isConnected) navigate("/");

    useEffect(() => {
        if (!isConnected) navigate("/") 
    })


    return (
        <>

            <Modal
                size={'small'}
                open={open}
                onClose={back}
                closeIcon={<Close />}
            >
                <Modal.Header>
                    <Header size="medium">Transfer</Header>
                    <p className='subtitle'>Send tokens to an account</p>
                </Modal.Header>

                <Modal.Content>
                    <Form>
                        <Field width={3} label="AMOUNT" placeholder="100" type="number" onChange={e => setAmount(Number(e.target.value))} />
                        <Field label="ADDRESS" type="address" onChange={e => setAddress(e.target.value)} />
                    </Form>
                </Modal.Content>
                <Modal.Actions>
                    <Button primary onClick={() => onSend()} loading={isTransfering}>
                        Send
                    </Button>
                    {error ? <p className="error">{error}</p> : null}

                </Modal.Actions>
            </Modal>

        </>
    )
}

export default Transfer
