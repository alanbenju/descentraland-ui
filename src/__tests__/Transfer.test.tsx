import { mount, configure } from 'enzyme';
import Transfer from '../components/Transfer/Transfer';
import { MemoryRouter as Router } from 'react-router-dom';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import * as React from 'react';

configure({ adapter: new Adapter() });

const mockUseLocationValue = {
    pathname: "/transfer",
    search: '',
    hash: '',
    state: null
}

jest.mock('react-router-dom', () => ({
    ...jest.requireActual("react-router-dom") as {},
    useNavigate: jest.fn().mockImplementation(() => {
        return mockUseLocationValue;
    })
}));


describe('<Transfer />', () => {
    describe('Transfer', () => {
        let component: any;
        let props: any;
        beforeAll(() => {
            props = {
                isConnected: true,
                error: "",
                onTransfer: jest.fn(),
                isTransfering: false,
            }
            component = mount(<Router><Transfer {...props} /></Router>);
        })

        it('Render Transfer connected', () => {
            expect(component.html()).toMatchSnapshot();
        });
        it('Click on transfer without address or amount', () => {
            let clickEvent = () => {
                component.find('button').simulate('click');
            };
            expect(clickEvent).not.toThrow();
            expect(props.onTransfer).not.toBeCalled()
        });
        it('Click on transfer with address and amount', () => {
            const setState = jest.fn();
            const useStateSpy = jest.spyOn(React, 'useState');
            useStateSpy.mockImplementationOnce(() => ["0x123", setState]);
            useStateSpy.mockImplementationOnce(() => [100, setState]);

            props = {
                isConnected: true,
                error: "",
                onTransfer: jest.fn(),
                isTransfering: false,
            }
            component = mount(<Transfer {...props} />);

            let clickEvent = () => {
                component.find('button').simulate('click');
            };
            expect(clickEvent).not.toThrow();
            expect(props.onTransfer).toBeCalled()

        })
    });

});
