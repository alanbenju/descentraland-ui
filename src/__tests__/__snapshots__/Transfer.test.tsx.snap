// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Transfer /> Transfer Render Transfer connected 1`] = `"<div class=\\"ui page modals dimmer transition visible active\\" style=\\"display: flex;\\"><div class=\\"ui small modal transition visible active\\"><div class=\\"dcl close \\"><div class=\\"close-icon\\"></div></div><div class=\\"header\\"><div class=\\"ui medium header\\">Transfer</div><p class=\\"subtitle\\">Send tokens to an account</p></div><div class=\\"content\\"><form class=\\"ui form\\"><div class=\\"dcl field has-label\\"><div class=\\"ui sub header\\">AMOUNT</div><div width=\\"3\\" class=\\"ui input\\"><input placeholder=\\"100\\" type=\\"number\\" value=\\"\\"></div><p class=\\"message\\">&nbsp;</p></div><div class=\\"dcl field address has-label\\"><div class=\\"ui sub header\\">ADDRESS</div><div class=\\"ui input\\"><input type=\\"text\\" value=\\"\\"></div><p class=\\"message\\">&nbsp;</p></div></form></div><div class=\\"actions\\"><button class=\\"ui primary button\\">Send</button></div></div></div>"`;
